compiler/rustc_mir_transform/src/match_branches.rs

```diff
@@ -515,3 +525,210 @@ impl<'tcx> SimplifyMatch<'tcx> for SimplifyToExp {
+        {
+            targets
+        }
+        _ => return None,
+    };
+    // We require that the possible target blocks all be distinct.
+    if !targets.is_distinct() {
+        return None;
+    }
+    if !bbs[targets.otherwise()].is_empty_unreachable() {
+        return None;
+    }
+    // Check that destinations are identical, and if not, then don't optimize this block.
+    let mut target_iter = targets.iter();
+    let first_terminator_kind = &bbs[target_iter.next().unwrap().1].terminator().kind;
+    if !target_iter
+        .all(|(_, other_target)| first_terminator_kind == &bbs[other_target].terminator().kind)
+    {
+        return None;
+    }
+
+    let borrowed_locals = borrowed_locals(body);
+    let mut live = None;
+
+    for (index, target_bb) in targets.iter() {
+        let stmts = &bbs[target_bb].statements;
+        if stmts.is_empty() {
+            return None;
+        }
+        if let [Statement { kind: StatementKind::Assign(box (place, rvalue)), .. }] =
+            bbs[target_bb].statements.as_slice()
+        {
+            let dest_ty = place.ty(body.local_decls(), tcx);
+            if dest_ty.ty != expected_src_ty.ty || dest_ty.variant_index.is_some() {
+                return None;
+            }
+            let ty::Adt(def, _) = dest_ty.ty.kind() else {
+                return None;
+            };
+            if expected_dest_place != *place {
+                return None;
+            }
+            match rvalue {
+                // Check if `_3 = const Foo::B` can be transformed to `_3 = copy *_1`.
+                Rvalue::Use(Operand::Constant(box constant))
+                    if let Const::Val(const_, ty) = constant.const_ =>
+                {
+                    let (ecx, op) =
+                        mk_eval_cx_for_const_val(tcx.at(constant.span), param_env, const_, ty)?;
+                    let variant = ecx.read_discriminant(&op).ok()?;
+                    if !def.variants()[variant].fields.is_empty() {
+                        return None;
+                    }
+                    let Discr { val, .. } = ty.discriminant_for_variant(tcx, variant)?;
+                    if val != index {
+                        return None;
+                    }
+                }
+                Rvalue::Use(Operand::Copy(src_place)) if *src_place == expected_src_place => {}
+                // Check if `_3 = Foo::B` can be transformed to `_3 = copy *_1`.
+                Rvalue::Aggregate(box AggregateKind::Adt(_, variant_index, _, _, None), fields)
+                    if fields.is_empty()
+                        && let Some(Discr { val, .. }) =
+                            expected_src_ty.ty.discriminant_for_variant(tcx, *variant_index)
+                        && val == index => {}
+                _ => return None,
+            }
+        } else {
+            // If the BB contains more than one statement, we have to check if these statements can be ignored.
```

Comment:
Can we merge both branches into the more general >1 statement case?

Reply:
Yes, but if all assignments are constants, there’s no need to check DSE.